### PR TITLE
Add the option to use serial mode.

### PR DESCRIPTION
**Pull request by Mauraza**
Add the option to use `SERIAL` instead of `PARALLEL` through a HTTP header.

Signed-off-by: mauraza <gibone@vmware.com>

### CODE REVIEW

**Review by gdelgadot — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -43,11 +43,12 @@ The above line is using the GitHub Action default input parameters. You can cust
...
 | pipeline               | This is the default JSON file that contains the VIB pipeline that will be executed.                                                                 | vib-pipeline.json    |
 | retry-count            | This is the default number of retries to do in case of failure reaching out to VIB.                                                                 | 3                    |
 | upload-artifacts       | This parameter specifies whether the GitHub Action will publish logs and reports as GitHub artifacts.                                               | true                 |
-| http-timeout           | This is the default number of seconds the GitHub Action waits for an HTTP timeout before failing.                                                   | 30000                |
+| verification_mode      | This paramerter allow to change the default parallel verification mode to serial.                                                                   | false                |
```
Comment (mpermar) [inline README.md:51]: To be consistent, this parameter should use a - not the _

File: action.yml [hunk 2]
```diff
@@ -33,6 +33,10 @@ inputs:
     description: 'Number of seconds the GitHub Action waits for an HTTP timeout before failing.'
     required: false
     default: '30000'
+  veridicaion-mode:
```
Comment (mpermar) [inline action.yml:36]: Typo here.

File: action.yml [hunk 3]
```diff
@@ -33,6 +33,10 @@ inputs:
...
+    description: 'VIB supports modifying the default parallel verification mode to serial.'
```
Comment (mpermar) [inline action.yml:37]:
```suggestion
    description: 'When specified, this GitHub Action will request VIB to execute the pipeline in the requested verification mode. Possible values: serial, parallel.'
```

File: src/main.ts [hunk 4]
```diff
@@ -30,14 +30,19 @@ export const cspClient = clients.newClient(
...
   {
     baseURL: `${process.env.VIB_PUBLIC_URL ? process.env.VIB_PUBLIC_URL : constants.DEFAULT_VIB_PUBLIC_URL}`,
     timeout: getNumberInput("http-timeout"),
-    headers: {
-      "Content-Type": "application/json",
-      "User-Agent": `vib-action/${userAgentVersion}`,
-    },
+    headers:
+      process.env.VERIFICATION_MODE === "true"
+        ? { ...defaultHeadersVibClient, "x-verification-mode": "SERIAL" }
```
Comment (mpermar) [inline src/main.ts:44]: Although HTTP headers are case insensitive, it's common to case them, i.e. suggesting to replace with `X-Verification-Mode`

Comment (mpermar) [inline src/main.ts:44]: Ok, this was a bit misleading. I thought (see comment above) the env var would define the verification mode. Not a boolean. If this is a boolean then the name of the variable should change to something like `serial-verification`. If not a boolean, then this line should use the value of the variable. 

I think my preference would be to let users specify the verification mode. There is nothing preventing us from adding new verification modes in the future, so the boolean approach is self-limiting.

**Review by mpermar — CHANGES_REQUESTED** [review]
Added some comments. In addition, I think this MR lacks some unit tests.

File: src/main.ts (hunk 4 above)
Comment (Mauraza) [reply src/main.ts:44]: A job appears `properties=[Property{headers=x-verification-mode=SERIAL}]}`

image

What is the correct value?

Comment (mpermar) [reply src/main.ts:44]: I mean doing this change:
```suggestion
        ? { ...defaultHeadersVibClient, "X-Verification-Mode": "SERIAL" }
```

File: action.yml [hunk 5]
```diff
@@ -33,10 +33,10 @@ inputs:
     description: 'Number of seconds the GitHub Action waits for an HTTP timeout before failing.'
     required: false
     default: '30000'
-  veridicaion-mode:
-    description: 'VIB supports modifying the default parallel verification mode to serial.'
+  verification-mode:
+    description: 'When specified, this GitHub Action will request VIB to execute the pipeline in the requested verification mode. Possible values: serial, parallel.'
```
Comment (mpermar) [inline action.yml:37]: My fault. 

```suggestion
    description: 'When specified, this GitHub Action will request VIB to execute the pipeline in the requested verification mode. Possible values: SERIAL, PARALLEL.'
```

File: __tests__/main.test.ts [hunk 6]
```diff
@@ -169,6 +169,27 @@ describe("VIB", () => {
       expect(config.pipeline).toEqual(constants.DEFAULT_PIPELINE)
     })
 
+    it("Default verification mode is used when not customized", async () => {
+      const config = await loadConfig()
+      console.log(config.verificationMode)
+      expect(config.verificationMode).toEqual(constants.DEFAULT_VERIFICATION_MODE)
+    })
+
+    it("Default verification mode is used when customized", async () => {
```
Comment (mpermar) [inline __tests__/main.test.ts:178]:
```suggestion
    it("Passed verification mode is used when customized", async () => {
```

File: src/main.ts [hunk 7]
```diff
@@ -382,6 +383,12 @@ export async function createPipeline(config: Config): Promise<string> {
   if (typeof process.env.VIB_PUBLIC_URL === "undefined") {
     core.setFailed("VIB_PUBLIC_URL environment variable not found.")
   }
+  const verificationModeValues = ["PARALLEL", "SERIAL"]
```
Comment (mpermar) [inline src/main.ts:386]: I think it is better if you move this to a constant in `constants.ts` and for consistency make it an enum.

File: src/main.ts [hunk 8]
```diff
@@ -382,6 +383,12 @@ export async function createPipeline(config: Config): Promise<string> {
...
+  core.info("verificationModeValues")
+  if (!verificationModeValues.includes(config.verificationMode)) {
+    core.setFailed("The value of Verification Mode is not valid, the default value will be used.")
```
Comment (mpermar) [inline src/main.ts:389]:
```suggestion
    core.setFailed(`The value ${config.verificationMode} for verification-mode is not valid, the default value will be used.`)
```

Comment (mpermar) [reply src/main.ts:389]: Needs to be a warning. If you do setFailed, that immediately stops the GitHub Action

File: __tests__/main.test.ts (hunk 6 above)
Comment (Mauraza) [reply __tests__/main.test.ts]: Done!

File: src/main.ts (hunk 8 above)
Comment (Mauraza) [reply src/main.ts]: Done

File: src/main.ts (hunk 7 above)
Comment (Mauraza) [reply src/main.ts]: Done!

**Review by mpermar — APPROVED** [review]
Great stuff.

**Pull request merged**